substrate/frame/support/src/generate_genesis_config.rs

```diff
@@ -705,6 +777,324 @@ mod test {
+				},
+				{
+					"t" : { "a": 32 },
+					"u" : { "a": 32, "s": { "a": 34, "b": 8} }
+				}
+			);
+		}
+	}
+
+	#[test]
+	fn test_generate_config_macro_struct_update() {
+		{
+			let s = S { x: 5 };
+			test!(TestStruct { s: S { ..s } }, { "s": { "x": 5 } });
+		}
+		{
+			mod nested {
+				use super::*;
+				pub fn function() -> S {
+					S { x: 5 }
+				}
+			}
+			test!(TestStruct { s: S { ..nested::function() } }, { "s": { "x": 5 } });
+		}
+		{
+			let s = nested_mod::InsideMod { a: 34, b: 8 };
+			let s1 = nested_mod::InsideMod { a: 34, b: 8 };
+			test!(
+				TestStruct {
+					t: nested_mod::InsideMod { ..s1 },
+					u: nested_mod::nested_mod2::nested_mod3::InsideMod3 {
+						s,
+						a: 32,
+					}
+				},
+				{
+					"t" : { "a": 34, "b": 8 },
+					"u" : { "a": 32, "s": { "a": 34, "b": 8} }
+				}
+			);
+		}
+		{
+			let i3 = nested_mod::nested_mod2::nested_mod3::InsideMod3 {
+				a: 1,
+				b: 2,
+				s: nested_mod::InsideMod { a: 55, b: 88 },
+			};
+			test!(
+				TestStruct {
+					t: nested_mod::InsideMod { a: 32 },
+					u: nested_mod::nested_mod2::nested_mod3::InsideMod3 {
+						a: 32,
+						..i3
+					}
+				},
+				{
+					"t" : { "a": 32 },
+					"u" : { "a": 32, "b": 2, "s": { "a": 55, "b": 88} }
+				}
+			);
+		}
+		{
+			let s = nested_mod::InsideMod { a: 34, b: 8 };
+			test!(
+				TestStruct {
+					t: nested_mod::InsideMod { a: 32 },
+					u: nested_mod::nested_mod2::nested_mod3::InsideMod3 {
+						a: 32,
+						s: nested_mod::InsideMod  {
+							b: 66,
+							..s
+						}
+					}
+				},
+				{
+					"t" : { "a": 32 },
+					"u" : { "a": 32, "s": { "a": 34, "b": 66} }
+				}
+			);
+		}
+		{
+			let s = nested_mod::InsideMod { a: 34, b: 8 };
+			test!(
+				TestStruct {
+					t: nested_mod::InsideMod { a: 32 },
+					u: nested_mod::nested_mod2::nested_mod3::InsideMod3 {
+						s: nested_mod::InsideMod  {
+							b: 66,
+							..s
+						},
+						a: 32
+					}
+				},
+				{
+					"t" : { "a": 32 },
+					"u" : { "a": 32, "s": { "a": 34, "b": 66} }
+				}
+			);
+		}
+	}
+
+	#[test]
+	fn test_generate_config_macro_with_execution_order() {
+		#[derive(Debug, Default, serde::Serialize, serde::Deserialize, PartialEq)]
+		struct X {
+			x: Vec<u32>,
+			x2: Vec<u32>,
+			y2: Y,
+		}
+		#[derive(Debug, Default, serde::Serialize, serde::Deserialize, PartialEq)]
+		struct Y {
+			y: Vec<u32>,
+		}
+		#[derive(Debug, Default, serde::Serialize, serde::Deserialize, PartialEq)]
+		struct Z {
+			a: u32,
+			x: X,
+			y: Y,
+		}
+		{
+			let v = vec![1, 2, 3];
+			test!(Z { a: 0, x: X { x: v },  }, {
+				"a": 0, "x": { "x": [1,2,3] }
+			});
+		}
+		{
+			let v = vec![1, 2, 3];
+			test!(Z { a: 3, x: X { x: v.clone() }, y: Y { y: v } }, {
+				"a": 3, "x": { "x": [1,2,3] }, "y": { "y": [1,2,3] }
+			});
+		}
+		{
+			let v = vec![1, 2, 3];
+			test!(Z { a: 3, x: X { y2: Y { y: v.clone() }, x: v.clone() }, y: Y { y: v } }, {
+				"a": 3, "x": { "x": [1,2,3], "y2":{ "y":[1,2,3] } }, "y": { "y": [1,2,3] }
+			});
```

Comment:
Shouldn't there be under the "x" key object a mapping between `"x2": []`?

Reply:
In this test we don't initialize `X::x2` struct field - so we expect that generated JSON blob will _not_ contain it.

Reply:
Hmm, I actually thought that default values will be visible in JSONs for uninitialized fields. Isn't that what's happening with genesis config presets?

Reply:
We want preset to contain only the fields that were explicitly initialized. We want preset to be a small patch containing only customized fields.

Otherwise we don't need macro, we could just instantiate `RuntimeGenesisConfig` struct  and simply serialize it to `json` (it works that way on master currently).

Reply:
Here is example for `development` of `asset_hub_rococo_runtime.compact.compressed.wasm` runtime:
<details><summary> preset containing only overridden fields</summary>

```json
{
  "balances": {
    "balances": [
      [
        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        1e+18
      ],
      [
        "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
        1e+18
      ],
      [
        "5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY",
        1e+18
      ],
      [
        "5HpG9w8EBLe5XCrbczpwq5TSXvedjrBGCwqxK1iQ7qUsSWFc",
        1e+18
      ]
    ]
  },
  "collatorSelection": {
    "candidacyBond": 53333328,
    "invulnerables": [
      "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
    ]
  },
  "parachainInfo": {
    "parachainId": 1000
  },
  "polkadotXcm": {
    "safeXcmVersion": 4
  },
  "session": {
    "keys": [
      [
        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        {
          "aura": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
        }
      ]
    ]
  }
}
```
</details>

<details><summary>full `RuntimeGenesisConfig` blob, containing defaults and overwritten fields (preset representation  currently on master)</summary>

```json
{
  "assets": {
    "accounts": [],
    "assets": [],
    "metadata": [],
    "nextAssetId": null
  },
  "aura": {
    "authorities": []
  },
  "auraExt": {},
  "balances": {
    "balances": [
      [
        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        1e+18
      ],
      [
        "5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty",
        1e+18
      ],
      [
        "5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY",
        1e+18
      ],
      [
        "5HpG9w8EBLe5XCrbczpwq5TSXvedjrBGCwqxK1iQ7qUsSWFc",
        1e+18
      ]
    ]
  },
  "collatorSelection": {
    "candidacyBond": 53333328,
    "desiredCandidates": 0,
    "invulnerables": [
      "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
    ]
  },
  "foreignAssets": {
    "accounts": [],
    "assets": [],
    "metadata": [],
    "nextAssetId": null
  },
  "parachainInfo": {
    "parachainId": 1000
  },
  "parachainSystem": {},
  "polkadotXcm": {
    "safeXcmVersion": 4
  },
  "poolAssets": {
    "accounts": [],
    "assets": [],
    "metadata": [],
    "nextAssetId": null
  },
  "session": {
    "keys": [
      [
        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY",
        {
          "aura": "5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY"
        }
      ]
    ],
    "nonAuthorityKeys": []
  },
  "system": {},
  "transactionPayment": {
    "multiplier": "1000000000000000000"
  }
}
```
</details>

Reply:
Thanks for the examples! How do we get the two different versions of the same preset? Does this mean the macro can be told when to convert to JSON only based on the initialized values vs when to convert the full struct schema to JSON (both initialized values and defaults)?

Reply:
Macro only generates patch/partial version of json blob.

If you ask about examples: for displaying presets I used two different runtimes - one from master, and one from my branch.

Reply:
I am reading things a bit differently, probably missing something. For example, parachain template genesis config starts with the partial set of fields and uses the macro, which adds the defaults and converts everything to JSON. At the end we should expect to have the full genesis config presets, right?

Reply:
Chatted offline with Michal who gave me a refresher of the flow of applying patches to the chain spec.
The idea of the macro is to return a JSON for the runtime config's initialized fields only, as if it is a diff from the default genesis state. This makes it clearer to what each preset would change, instead of returning the full config as before.

Resolving this!